pkg/cloudsources/paladin/client.go

```diff
@@ -0,0 +1,119 @@
+package paladin
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/httputil"
+	"github.com/stackrox/rox/pkg/httputil/proxy"
+	"github.com/stackrox/rox/pkg/utils"
+	"github.com/stackrox/rox/pkg/version"
+)
+
+// AssetsResponse holds the response returned by the Paladin Cloud API.
+type AssetsResponse struct {
+	Assets []Asset `json:"assets,omitempty"`
+}
+
+// Asset holds the asset as returned by the Paladin Cloud API.
+type Asset struct {
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	Type   string `json:"type"`
+	Source string `json:"source"`
+	Region string `json:"region"`
+}
+
+// Client can be used to interact with the Paladin Cloud API.
+type Client struct {
+	httpClient *http.Client
+	endpoint   string
+}
+
+// paladinTransportWrapper adds auth information to the underlying transport as well as the user agent.
+type paladinTransportWrapper struct {
+	baseTransport http.RoundTripper
+	token         string
+	acsVersion    string
+}
+
+func (p *paladinTransportWrapper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// The Paladin Cloud API expected the Authorization header in the format '"Authorization:" <TOKEN>'
+	// instead of e.g. Bearer token format.
+	req.Header.Set("Authorization", p.token)
+
+	req.Header.Set("User-Agent", "RHACS/"+p.acsVersion)
+	return p.baseTransport.RoundTrip(req)
+}
+
+// NewClient creates a client to interact with Paladin Cloud APIs.
+func NewClient(cfg *storage.CloudSource) *Client {
+	httpClient := &http.Client{
+		Transport: &paladinTransportWrapper{
+			baseTransport: proxy.RoundTripper(),
+			token:         cfg.GetCredentials().GetSecret(),
+			acsVersion:    version.GetMainVersion(),
+		},
+		Timeout: time.Minute,
+	}
+
+	return &Client{
+		httpClient: httpClient,
+		endpoint:   cfg.GetPaladinCloud().GetEndpoint(),
+	}
+}
+
+// GetAssets returns the discovered assets from Paladin Cloud.
+func (c *Client) GetAssets(ctx context.Context) (*AssetsResponse, error) {
+	var assets AssetsResponse
+
+	if err := c.sendRequest(ctx, http.MethodGet, "/v2/assets", &assets); err != nil {
+		return nil, errors.Wrap(err, "retrieving assets")
+	}
+
+	return &assets, nil
+}
+
+func (c *Client) sendRequest(ctx context.Context, method string, apiPath string, response interface{}) error {
+	req, err := http.NewRequestWithContext(ctx, method, path.Join(c.endpoint, apiPath), nil)
```

Comment:
Which one is better path or url?
```suggestion
	req, err := http.NewRequestWithContext(ctx, method, url.JoinPath(c.endpoint, apiPath), nil)
```
https://pkg.go.dev/net/url#JoinPath